pkg/controller/userdatasecret/userdatasecret_controller.go

```diff
@@ -0,0 +1,217 @@
+				return true
+			}
+			return false
+		},
+	}
+
+	// Watch for changes to primary resource UserDataSecret
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{ObjectMeta: meta.ObjectMeta{Namespace: "openshift-machine-api"}}}, &handler.EnqueueRequestForObject{}, predicateFilter)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// blank assignment to verify that ReconcileUserDataSecret implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileUserDataSecret{}
+
+// ReconcileUserDataSecret reconciles a UserDataSecret object
+type ReconcileUserDataSecret struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	// signer is a signer created from the user's private key
+	signer ssh.Signer
+}
+
+// Reconcile reads that state of the cluster for a UserDataSecret object and makes changes based on the state read
+// and what is in the UserDataSecret.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileUserDataSecret) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling UserDataSecret")
+
+	userData, err := r.createUserDataSecret()
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "error creating user data secret")
+
+	}
+	// Fetch UserData instance
+	err = r.client.Get(context.TODO(), kubeTypes.NamespacedName{Name: userData.Name, Namespace: userData.Namespace}, &corev1.Secret{})
+	if err != nil && k8sapierrors.IsNotFound(err) {
+		// Secret is deleted
+		reqLogger.Info("Creating a new Secret", "Secret.Namespace", userData.Namespace, "Secret.Name", userData.Name)
+		err = r.client.Create(context.TODO(), userData)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// Secret created successfully - don't requeue
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{}, err
+	} else {
+		// secret is updated
+		reqLogger.Info("Updating the secret", "Secret.Namespace", userData.Namespace, "Secret.Name", userData.Name)
+		err = r.client.Update(context.TODO(), userData)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// Secret updated successfully
+		return reconcile.Result{}, nil
+	}
+}
+
+// createUserDataSecret creates a secret 'windows-user-data' in 'openshift-machine-api'
+// namespace. This secret will be used to inject cloud provider user data for creating
+// windows machines
+func (r *ReconcileUserDataSecret) createUserDataSecret() (*corev1.Secret, error) {
```

Comment:
This isnt actually creating a userdata secret, it is returning a secret that can be created.

Reply:
You can change the name to reflect the functionality, or change the function to reflect the name.

Reply:
Please fix the doc too. `construct` or `generate` might be more appropriate here.